Finish adding PHPUnit 10 support and test with PHP 8.2 and 8.3

Krinkle opened the pull request:
Finish adding PHPUnit 10 support

### Issues with PHPUnit 9.6 on PHP 7.4

```
1) OckCyp\CoversValidator\Tests\Command\ValidateCommandTest::testPrintsConfigFileUsed
RuntimeException: Could not find PHPUnit TestSuiteMapper class
```

Restore the PHPUnit 9.3 and 9.5 logic from the master branch.

In PHPUnit 10, TestSuiteMapper was renamed to the same name as 9.3,
but takes different args. This means the map() call needs to become
part of the conditional. It also requires the ConfigurationRegistry
singleton to be populated, because otherwise:

> $ vendor/bin/covers-validator
> CoversValidator 1.6.0
> PHP Fatal error:  Uncaught AssertionError: assert(self::$instance instanceof Configuration)
>   in /vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php:93
> Stack trace:
> /vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php(93)
> /vendor/phpunit/phpunit/src/Framework/TestBuilder.php(128): Registry::get
> /vendor/phpunit/phpunit/src/Framework/TestBuilder.php(63): TestBuilder->configureTestCase
> /vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): …
> /vendor/phpunit/phpunit/src/TextUI/Configuration/Xml/TestSuiteMapper.php(102): …
> /vendor/ockcyp/covers-validator/src/Model/TestCollection.php(50): TestSuiteMapper->map

### Issues with PHPUnit 10 on PHP 8

```
…\InputHandlerTest::testHandlesInputWithNoBootstrap
Error: Call to undefined method …\InputHandlerTest::prophesize()
```

https://github.com/sebastianbergmann/phpunit/blob/9.6.26/DEPRECATIONS.md#test-double-api-1
https://phpunit.de/announcements/phpunit-10.html
https://github.com/sebastianbergmann/phpunit/blob/10.0.0/ChangeLog-10.0.md
https://github.com/sebastianbergmann/phpunit/issues/4142

```
…\ConfigLoaderTest::testLoadsConfig
Error: Call to undefined method …\ConfigLoaderTest::preTestRun()
```

Restore previous logic for now, which seemed to work fine.

```
…\TestSuiteLoaderTest::testLoadsSuitePHPUnit9
Error: Class "PHPUnit\TextUI\Configuration\Loader" not found
```

Condition checked an unused PHPUnit9 classes, which happens to exist
on PHPUnit10 as well. The class that is actually used (Loader) was
renamed again.

Update the condition, and write a new test for PHPUnit 10. Note that
TestCase::getName() was removed as well, so update that to use
TestName::name() instead, same as in ValidateCommand already.

In addition, switch the condition from a class check to a version
check because otherwise:

> PHPUnit 9.6.26
> TestSuiteLoaderTest::testLoadsSuitePHPUnit10
> Error: Call to undefined method OneTestCoveringExistingTwoClassTest::name()

... because the PHPUnit 10+ class was first introduced in PHPUnit 9.3.

I realized the same logic exists in ConfigLoader::loadConfig, which
seems worth re-using (and testing) instead, so I've simplified the
test to use that instead.

```
1) OckCyp\CoversValidator\Tests\Loader\FileLoaderTest::testLoadsFile
Failed asserting that false is true.

2) OckCyp\CoversValidator\Tests\Loader\FileLoaderTest::testLoadsFile2
Failed asserting that false is true.
```

In PHPUnit 10, the FileLoader::checkAndLoad utility was removed,
and with it support for setting global variables from there.
The equivalent is simply include_once and that is what PHPUnit 10+
does as well.

The example test here, however, expected PHPUnit 9-specific behavior.
Skip it under PHPUnit 10+ and write a new test that defines a function
instead, which does not rely on emulating global scope for vars.

```
1) …\OneTestWithProviderCoveringNonExistentClassTest::testDummyTest
Data Provider method provideDummyTest() is not static

2) …\OneTestWithProviderCoveringExistingClassTest::testDummyTest
Data Provider method provideDummyTest() is not static
```

Fixed.

### Issues with PHPUnit 10 coverage build

```
Error: No filter is configured, code coverage will not be processed
```

Migrate phpunit.xml.dist, and set the new one for the coverage
build only since it requires PHPUnit 10. It only runs on one
version anyway.

```
Run vendor/bin/php-coveralls
COVERALLS_REPO_TOKEN=""
Error: Process completed with exit code 1.
```

Skip upload step on forks where secret tokens don't apply.
Keep the rest of the coverage build enabled, so that it can
be tested.

Ref https://github.com/oradwell/covers-validator/issues/44.

Krinkle committed:
Finish adding PHPUnit 10 support and test with PHP 8.2 and 8.3

=== Issues with PHPUnit 9.6 on PHP 7.4 ===

```
1) OckCyp\CoversValidator\Tests\Command\ValidateCommandTest::testPrintsConfigFileUsed
RuntimeException: Could not find PHPUnit TestSuiteMapper class
```

Restore the PHPUnit 9.3 and 9.5 logic from the master branch.

In PHPUnit 10, TestSuiteMapper was renamed to the same name as 9.3,
but takes different args. This means the map() call needs to become
part of the conditional. It also requires the ConfigurationRegistry
singleton to be populated, because otherwise:

> $ vendor/bin/covers-validator
> CoversValidator 1.6.0
> PHP Fatal error:  Uncaught AssertionError: assert(self::$instance instanceof Configuration)
>   in /vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php:93
> Stack trace:
> /vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php(93)
> /vendor/phpunit/phpunit/src/Framework/TestBuilder.php(128): Registry::get
> /vendor/phpunit/phpunit/src/Framework/TestBuilder.php(63): TestBuilder->configureTestCase
> /vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): …
> /vendor/phpunit/phpunit/src/TextUI/Configuration/Xml/TestSuiteMapper.php(102): …
> /vendor/ockcyp/covers-validator/src/Model/TestCollection.php(50): TestSuiteMapper->map

=== Issues with PHPUnit 10 on PHP 8 ===

```
…\InputHandlerTest::testHandlesInputWithNoBootstrap
Error: Call to undefined method …\InputHandlerTest::prophesize()
```

https://github.com/sebastianbergmann/phpunit/blob/9.6.26/DEPRECATIONS.md#test-double-api-1
https://phpunit.de/announcements/phpunit-10.html
https://github.com/sebastianbergmann/phpunit/blob/10.0.0/ChangeLog-10.0.md
https://github.com/sebastianbergmann/phpunit/issues/4142

```
…\ConfigLoaderTest::testLoadsConfig
Error: Call to undefined method …\ConfigLoaderTest::preTestRun()
```

Restore previous logic for now, which seemed to work fine.

```
…\TestSuiteLoaderTest::testLoadsSuitePHPUnit9
Error: Class "PHPUnit\TextUI\Configuration\Loader" not found
```

Condition checked an unused PHPUnit9 classes, which happens to exist
on PHPUnit10 as well. The class that is actually used (Loader) was
renamed again.

Update the condition, and write a new test for PHPUnit 10. Note that
TestCase::getName() was removed as well, so update that to use
TestName::name() instead, same as in ValidateCommand already.

In addition, switch the condition from a class check to a version
check because otherwise:

> PHPUnit 9.6.26
> TestSuiteLoaderTest::testLoadsSuitePHPUnit10
> Error: Call to undefined method OneTestCoveringExistingTwoClassTest::name()

... because the PHPUnit 10+ class was first introduced in PHPUnit 9.3.

I realized the same logic exists in ConfigLoader::loadConfig, which
seems worth re-using (and testing) instead, so I've simplified the
test to use that instead.

```
1) OckCyp\CoversValidator\Tests\Loader\FileLoaderTest::testLoadsFile
Failed asserting that false is true.

2) OckCyp\CoversValidator\Tests\Loader\FileLoaderTest::testLoadsFile2
Failed asserting that false is true.
```

In PHPUnit 10, the FileLoader::checkAndLoad utility was removed,
and with it support for setting global variables from there.
The equivalent is simply include_once and that is what PHPUnit 10+
does as well.

The example test here, however, expected PHPUnit 9-specific behavior.
Skip it under PHPUnit 10+ and write a new test that defines a function
instead, which does not rely on emulating global scope for vars.

```
1) …\OneTestWithProviderCoveringNonExistentClassTest::testDummyTest
Data Provider method provideDummyTest() is not static

2) …\OneTestWithProviderCoveringExistingClassTest::testDummyTest
Data Provider method provideDummyTest() is not static
```

Fixed.

=== Issues with PHPUnit 10 coverage build ===

```
Error: No filter is configured, code coverage will not be processed
```

Migrate phpunit.xml.dist, and set the new one for the coverage
build only since it requires PHPUnit 10. It only runs on one
version anyway.

```
Run vendor/bin/php-coveralls
COVERALLS_REPO_TOKEN=""
Error: Process completed with exit code 1.
```

Skip upload step on forks where secret tokens don't apply.
Keep the rest of the coverage build enabled, so that it can
be tested.

Ref https://github.com/oradwell/covers-validator/issues/44.

Krinkle commented:
@oradwell It is now passing for me at: https://github.com/Krinkle/covers-validator/actions/runs/17690148302

oradwell commented:
I've had a quick look. Great work! Thanks so much for your efforts. I'll take a deeper look and move it forward.

Krinkle commented:
@oradwell Thanks. While CI is passing here, when I tried to use this today in https://github.com/wikimedia/Minify, I did run into a problem.

```
diff --git a/composer.json b/composer.json
index ce61a84b63..baf08ff49e 100644
--- a/composer.json
+++ b/composer.json
@@ -36,15 +36,21 @@
 		"ext-mbstring": "*",
 		"pear/net_url2": "~2.2.2"
 	},
+	"repositories": [
+		{
+			"type": "git",
+			"url": "https://github.com/Krinkle/covers-validator.git"
+		}
+	],
 	"require-dev": {
 		"mck89/peast": "^1.16",
 		"mediawiki/mediawiki-codesniffer": "48.0.0",
 		"mediawiki/mediawiki-phan-config": "0.17.0",
 		"mediawiki/minus-x": "1.1.3",
-		"ockcyp/covers-validator": "1.6.0",
+		"ockcyp/covers-validator": "dev-phpunit-10",
 		"php-parallel-lint/php-console-highlighter": "1.0.0",
 		"php-parallel-lint/php-parallel-lint": "1.4.0",
-		"phpunit/phpunit": "9.6.21",
+		"phpunit/phpunit": "10.x",
 		"wikimedia/testing-access-wrapper": "~3.0.0"
 	},
 	"scripts": {
```
```
$ php --version
PHP 8.3.24 …
```
```
$ composer update
…
  - Upgrading phpunit/php-code-coverage (9.2.32 => 10.1.16): Extracting archive
  - Upgrading phpunit/phpunit (9.6.21 => 10.5.54): Extracting archive
…
```
```
$ vendor/bin/covers-validator 
CoversValidator 1.6.0
PHP Fatal error:  Uncaught AssertionError: assert(self::$instance instanceof Configuration) in /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php:93
Stack trace:
#0 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php(93): assert(false, 'assert(self::$i...')
#1 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestBuilder.php(128): PHPUnit\TextUI\Configuration\Registry::get()
#2 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestBuilder.php(63): PHPUnit\Framework\TestBuilder->configureTestCase(Object(CSSMinTest), false, NULL, false, Array)
#3 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): PHPUnit\Framework\TestBuilder->build(Object(ReflectionClass), 'testAlias')
#4 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestSuite.php(139): PHPUnit\Framework\TestSuite->addTestMethod(Object(ReflectionClass), Object(ReflectionMethod))
#5 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestSuite.php(231): PHPUnit\Framework\TestSuite::fromClassReflector(Object(ReflectionClass))
#6 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestSuite.php(250): PHPUnit\Framework\TestSuite->addTestSuite(Object(ReflectionClass))
#7 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/Framework/TestSuite.php(269): PHPUnit\Framework\TestSuite->addTestFile('/Users/krinkle/...')
#8 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/TextUI/Configuration/Xml/TestSuiteMapper.php(102): PHPUnit\Framework\TestSuite->addTestFiles(Array)
#9 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/ockcyp/covers-validator/src/Model/TestCollection.php(50): PHPUnit\TextUI\XmlConfiguration\TestSuiteMapper->map('/Users/krinkle/...', Object(PHPUnit\TextUI\Configuration\TestSuiteCollection), '', '')
#10 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/ockcyp/covers-validator/src/Loader/TestSuiteLoader.php(15): OckCyp\CoversValidator\Model\TestCollection->__construct(Object(OckCyp\CoversValidator\Model\ConfigurationHolder))
#11 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/ockcyp/covers-validator/src/Command/ValidateCommand.php(60): OckCyp\CoversValidator\Loader\TestSuiteLoader::loadSuite(Object(OckCyp\CoversValidator\Model\ConfigurationHolder))
#12 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/symfony/console/Command/Command.php(326): OckCyp\CoversValidator\Command\ValidateCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand(Object(OckCyp\CoversValidator\Command\ValidateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/ockcyp/covers-validator/covers-validator(16): Symfony\Component\Console\Application->run()
#17 /Users/krinkle/Development/mediawiki-libs-Minify/vendor/bin/covers-validator(119): include('/Users/krinkle/...')
#18 {main}
  thrown in /Users/krinkle/Development/mediawiki-libs-Minify/vendor/phpunit/phpunit/src/TextUI/Configuration/Registry.php on line 93
```

I'm giving it a bit more energy to see if I can solve this one too. Looks like something somewhere still isn't quite right.

Krinkle commented:
It seems a the issues stems from a difference in `TestSuiteMapper` between 9.x and 10.x.

In 9.x, the mapper would call `TestSuite->addTestSuite`https://github.com/sebastianbergmann/phpunit/blob/9.6/src/Framework/TestSuite.php#L309 , which merely creates a ReflectionClass for the test case subclass, and then moves on.

In 10.x, this step is outsourced to `TestSuite->fromClassReflector`, which goes deeper and starts building things in preparation for test execution at https://github.com/sebastianbergmann/phpunit/blob/10.5/src/Framework/TestSuite.php#L503 via `TestSuite->addTestMethod`, which eventually reaches `TestBuilder->configureTestCase` https://github.com/sebastianbergmann/phpunit/blob/10.5/src/Framework/TestBuilder.php#L111 and expects to the `PHPUnit\TextUI\Configuration\Registry` singleton (as ConfigurationRegistry) to be non-empty. That singleton is tricky to initialise, because it has no clean public setter. It has a loadFrom method, which takes a custom format we can't easily write (the saveTo method requires having the singleton be populated already), or `::init` which is presumably how PHPUnit itself sets this but that requires both a CliConfiguration and an XmlConfiguration, whereas we only have an XmlConfiguration.

On the bright side, while CliConfiguration is a mess to create manually, it is easy to create a generate a complete null version via `(new \PHPUnit\TextUI\CliArguments\Builder)->fromParameters([]),` which ConfigurationRegistry::init can then do a no-op merge on with our XmlConfiguration.

Doing that seems to work. I ran it on our wikimedia/minify project and got a clean build, both in the positive and an expected well-formatted error and excit code when adding some intentional mistakes.

```diff
+		"ockcyp/covers-validator": "dev-phpunit-10#f58bd44cf4",
```
```
CoversValidator 1.6.0

Validation complete. All @covers tags are valid.
```
```
$ composer test
…
CoversValidator 1.6.0

Invalid - IndexMapTest::testEmptyMap
Invalid - IndexMapTest::testOutputFile
Invalid - IndexMapTest::testOneSection
Invalid - IndexMapTest::testTwoSections

There were 4 test(s) with invalid @covers tags.
Script covers-validator handling the test event returned with error code 1
```

oradwell commented on tests/Handler/InputHandlerTest.php:
```diff
@@ -5,18 +5,39 @@
     public function testHandlesInputWithNoBootstrap()
     {
-        $input = $this->prophesize(
+        $input = $this->prophet->prophesize(
```
is it possible to have a method that adapts to PHP 7.1 for this one? all it needs to do is to use `$this->prophesize()` vs `$this->prophet->prophesize()`

oradwell commented on tests/Handler/InputHandlerTest.php:
```diff
@@ -5,18 +5,39 @@
  */
 class InputHandlerTest extends BaseTestCase
 {
+    private $prophet;
```
```suggestion
    /**
     * @var Prophet|null
     */
    private $prophet;
```

oradwell reviewed (changes requested):
great work here! just a couple of comments. let me know what you think

oradwell commented:
Given that the CI passed but failed on your project. Do you think we should have an integration test covering this case?

Krinkle commented on tests/Handler/InputHandlerTest.php:
```diff
@@ -5,18 +5,39 @@
+    protected function setUp(): void
+    {
+        if (!class_exists(Prophet::class)) {
+            $this->markTestSkipped('Skip Prophet-based tests on PHP 7.1');
```
I misunderstood what's going on. I noticed the CI setup for this repo did not install Prophecy on the  PHP 7.1 build. I wrongly assumed this was because it's old and not supported there. I further more wrongy assumed that PHPUnit is detecting the package's optional presence and using it and otherwise no-op'ing this.

Looking at it more closely, that's not the case. The PHP 7.1 build ends up installing PHPUnit 7.x, which actually installs the same prophecy package directly, as a dependency. So it is fully present in all builds. Let me remove this and see if it passes.

Krinkle committed:
tests: Run ValidateCommandTest cases as sub processes

This would have caught the PHPUnit 10 breakage due to
an unset ConfigurationRegistry singleton. This was missed in the
integration, because it invoked ValidateCommand in the same process
as part of an already-configured `phpunit` program, and thus the
global state is already fine.

If you comment out these three lines in src/Mode/TestCollection.php,
the previous tests would pass, but the new tests fail:

```php
 \PHPUnit\TextUI\Configuration\Registry::init(
    (new \PHPUnit\TextUI\CliArguments\Builder())->fromParameters([]),
    $configuration
);
```

Follows-up https://github.com/oradwell/covers-validator/pull/48.
